src/types/component.ts

```diff
@@ -31,3 +33,230 @@ export enum ComponentType {
   Calculate = 700,
   Confirmation = 725,
 }
+
+export function component(type: ComponentType): Component {
+  return ComponentSchema[ComponentType[type] as keyof typeof ComponentSchema];
+}
+
+export type Component = (typeof ComponentSchema)[keyof typeof ComponentSchema];
+
+export type ComponentSchemaEntry = {
+  type: ComponentType;
+  properties?: {
+    reads?: readonly PropertyDefinition[];
+    writes?: readonly PropertyDefinition[];
+  };
+  branching?: BranchingRule;
+  placementRestriction?: {
+    previous?: readonly ComponentType[]; // previous node must be one of these types
+    next?: readonly ComponentType[]; // next node must be one of these types
+  };
+};
```

Comment:
I'm letting the types drive the design here a little (i.e. some of these fields aren't used yet). 
Here, the functionality of components are declared as a data structure. This means that logic functions can query the intended behaviour of the component explicitly. 
For example, each component has branching rules which define what nodes can be nested and what kind of collection the nesting represents. This allows for traversal without having to special case particular component types in the logic functions.